futures-util/src/future/abortable.rs

```diff
@@ -0,0 +1,131 @@
+use futures_core::future::Future;
+use futures_core::task::{self, AtomicWaker, Poll};
+use std::marker::Unpin;
+use std::mem::PinMut;
+use std::sync::Arc;
+use std::sync::atomic::{AtomicBool, Ordering};
+
+/// A future which can be remotely short-circuited using an `AbortHandle`.
+#[derive(Debug, Clone)]
+#[must_use = "futures do nothing unless polled"]
+pub struct Abortable<T> {
+    future: T,
+    inner: Arc<AbortInner>,
+}
+
+impl<T: Unpin> Unpin for Abortable<T> {}
+
+impl<T> Abortable<T> where T: Future {
+    /// Creates a new `Abortable` future using an existing `AbortRegistration`.
+    /// `AbortRegistration`s can be acquired through `AbortHandle::new`.
+    ///
+    /// When `abort` is called on the handle tied to `reg` or if `abort` has
+    /// already been called, the future will complete immediately without making
+    /// any further progress.
+    pub fn new(future: T, reg: AbortRegistration) -> Self {
+        Abortable {
+            future: future,
+            inner: reg.inner,
+        }
+    }
+
+    unsafe_pinned!(future: T);
```

Comment:
I think we elsewhere always list pinned projections at the top

Reply:
Done.